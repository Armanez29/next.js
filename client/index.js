import { createElement } from 'react'
import ReactDOM from 'react-dom'
import HeadManager from './head-manager'
import { rehydrate } from '../lib/css'
import { createRouter } from '../lib/router'
import App from '../lib/app'
import evalScript from '../lib/eval-script'

const {
  __NEXT_DATA__: {
    component,
    errorComponent,
    props,
    ids,
    err,
    pathname,
    query
  }
} = window

const Component = evalScript(component).default
const ErrorComponent = evalScript(errorComponent).default
let lastAppProps

export const router = createRouter(pathname, query, {
  Component,
  ErrorComponent,
  err
})

const headManager = new HeadManager()
const container = document.getElementById('__next')

export default (onError) => {
  if (ids && ids.length) rehydrate(ids)

  router.subscribe(({ Component, props, err }) => {
    render({ Component, props, err }, onError)
  })

  render({ Component, props, err }, onError)
}

export function render (props, onError = renderErrorComponent) {
  try {
    doRender(props)
  } catch (err) {
    onError(err)
  }
}

async function renderErrorComponent (err) {
  const { pathname, query } = router
  const props = await getInitialProps(ErrorComponent, { err, pathname, query })
  doRender({ Component: ErrorComponent, props, err })
}

function doRender ({ Component, props, err }) {
  Component = Component || lastAppProps.Component
  props = props || lastAppProps.props

  const appProps = { Component, props, err, router, headManager }
  lastAppProps = appProps
  ReactDOM.render(createElement(App, appProps), container)
}

function getInitialProps (Component, ctx) {
  return Component.getInitialProps ? Component.getInitialProps(ctx) : {}
}
