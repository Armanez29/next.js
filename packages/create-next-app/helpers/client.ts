/* eslint-disable import/no-extraneous-dependencies */
import got from 'got'
import { HttpsProxyAgent } from 'hpagent'

function getAgent() {
  if (process.env.https_proxy) {
    return {
      https: new HttpsProxyAgent({
        proxy: process.env.https_proxy,
        rejectUnauthorized: false,
      }),
    }
  }
}

export default got.extend({ agent: getAgent() })
