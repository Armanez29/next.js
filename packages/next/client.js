module.exports = require('./dist/api/client')
