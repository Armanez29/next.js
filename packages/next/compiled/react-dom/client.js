/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
/******/ 	var __webpack_modules__ = ({

/***/ 174:
/***/ ((module) => {

module.exports = require("next/dist/compiled/react-dom");

/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __nccwpck_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		var threw = true;
/******/ 		try {
/******/ 			__webpack_modules__[moduleId](module, module.exports, __nccwpck_require__);
/******/ 			threw = false;
/******/ 		} finally {
/******/ 			if(threw) delete __webpack_module_cache__[moduleId];
/******/ 		}
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/compat */
/******/ 	
/******/ 	if (typeof __nccwpck_require__ !== 'undefined') __nccwpck_require__.ab = __dirname + "/";
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
var exports = __webpack_exports__;


var m = __nccwpck_require__(174);
if (process.env.NODE_ENV === 'production') {
  exports.createRoot = m.createRoot;
  exports.hydrateRoot = m.hydrateRoot;
} else {
  var i = m.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
  exports.createRoot = function(c, o) {
    i.usingClientEntryPoint = true;
    try {
      return m.createRoot(c, o);
    } finally {
      i.usingClientEntryPoint = false;
    }
  };
  exports.hydrateRoot = function(c, h, o) {
    i.usingClientEntryPoint = true;
    try {
      return m.hydrateRoot(c, h, o);
    } finally {
      i.usingClientEntryPoint = false;
    }
  };
}

})();

module.exports = __webpack_exports__;
/******/ })()
;