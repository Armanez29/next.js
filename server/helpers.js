import { join } from 'path'
import fs from 'mz/fs'
import glob from 'glob-promise'

export async function getParamRoutes (base) {
  let files = await glob(join(base, '**'))

  const result = await Promise.all(files.map(f => isFile(f)))
  files = result.reduce((_files, isFile, i) => {
    return isFile ? [..._files, files[i]] : _files
  })

  files = files.filter(f => f.indexOf('}') > 0)
  files = files.map(f => f.replace(base, ''))
  return files
}

export async function isFile (p) {
  let stat
  try {
    stat = await fs.stat(p)
  } catch (err) {
    if (err.code === 'ENOENT') return false
    throw err
  }
  return stat.isFile() || stat.isFIFO()
}

export function toParts (route) {
  return route.replace(/^\/|\.js$/g, '').split('/')
}

export function compareRoute (parts, segments) {
  // parts:    ['user', '{id}', 'action']
  // segments: ['user', '1234', 'action']
  if (parts.length === segments.length) {
    return parts.map((p, i) => p.startsWith('{') || segments[i] === parts[i]).every(ok => ok)
  }
  return false
}
