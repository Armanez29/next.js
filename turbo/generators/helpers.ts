import type { PlopTypes } from '@turbo/gen'

export function toFileName(str: string) {
  return str.toLowerCase().replace(/ /g, '-')
}

export function init(plop: PlopTypes.NodePlopAPI): void {
  plop.setHelper('toFileName', toFileName)
}
