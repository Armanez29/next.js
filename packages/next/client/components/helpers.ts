import type { CacheNode } from '../../shared/lib/app-router-context'
import type { FlightRouterState } from '../../server/app-render'
import { fetchServerResponse } from './app-router.client'

export const createOptimisticTree = (
  existingState: any,
  url: URL,
  // /dashboard/integrations ['', 'dashboard', 'integrations', 'page']
  segments: string[],
  existingFlightRouterState: FlightRouterState | null,
  childNodes: CacheNode['childNodes'] | null,
  isFirstSegment: boolean,
  parentFetchingData: boolean,
  initialMap = new Map()
): {
  childNodes: CacheNode['childNodes']
  tree: FlightRouterState
} => {
  const [existingSegment, existingParallelRoutes] =
    existingFlightRouterState || [null, {}]
  const segment = segments[0]
  const isLastSegment = segments.length === 1

  let parallelRoutes: FlightRouterState[1] = {}
  if (existingSegment === segment) {
    parallelRoutes = existingParallelRoutes
  }

  if (childNodes) {
    for (const [key, value] of childNodes) {
      initialMap.set(key, {
        subTreeData: value.subTreeData,
        childNodes: new Map(),
      })
    }
  }

  let startedFetch = false
  if (!isFirstSegment && !parentFetchingData) {
    // Seed the cache with missing item
    if (childNodes && !childNodes.has(segment)) {
      startedFetch = true
      console.log('KICKING OFF DATA FETCH IN PUSH', {
        segment,
        url,
      })

      initialMap.set(segment, {
        data: fetchServerResponse(url, existingState.tree),
        subTreeData: null,
        childNodes: new Map(),
      })
    }
  }

  let childTree
  if (!isLastSegment) {
    const passedChildNodes = isFirstSegment
      ? childNodes
      : childNodes && childNodes.has(segment)
      ? childNodes.get(segment)!.childNodes
      : null
    const childItem = createOptimisticTree(
      existingState,
      url,
      segments.slice(1),
      parallelRoutes ? parallelRoutes.children : null,
      passedChildNodes,
      false,
      parentFetchingData || startedFetch,
      isFirstSegment ? initialMap : initialMap.get(segment)!.childNodes
    )

    childTree = childItem.tree
  }

  const result: FlightRouterState = [
    segment,
    {
      ...parallelRoutes,
      ...(childTree ? { children: childTree } : {}),
    },
  ]

  // Add url into the tree
  if (isFirstSegment) {
    result[2] = url.pathname + url.search
  }

  return {
    childNodes: initialMap,
    tree: result,
  }
}
