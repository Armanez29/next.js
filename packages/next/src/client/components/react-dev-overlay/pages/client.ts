import * as Bus from './bus'
import { parseStack } from '../internal/helpers/parse-stack'
import { parseComponentStack } from '../internal/helpers/parse-component-stack'
import {
  hydrationErrorState,
  storeHydrationErrorStateFromConsoleArgs,
} from '../internal/helpers/hydration-error-info'
import {
  ACTION_BEFORE_REFRESH,
  ACTION_BUILD_ERROR,
  ACTION_BUILD_OK,
  ACTION_REFRESH,
  ACTION_UNHANDLED_ERROR,
  ACTION_UNHANDLED_REJECTION,
  ACTION_VERSION_INFO,
} from '../shared'
import type { VersionInfo } from '../../../../server/dev/parse-version-info'
import { attachHydrationErrorState } from '../internal/helpers/attach-hydration-error-state'
import {
  patchConsoleError,
  unpatchConsoleError,
} from '../../globals/intercept-console-error'
import { ConsoleError } from '../internal/helpers/console-error'
import { formatConsoleArgs } from '../../../lib/console'
import isError from '../../../../lib/is-error'

let isRegistered = false
let stackTraceLimit: number | undefined = undefined

function handleClientError(originError: unknown, consoleErrorArgs: unknown[]) {
  let error: Error
  if (!originError || !isError(originError)) {
    // If it's not an error, format the args into an error
    const formattedErrorMessage = formatConsoleArgs(consoleErrorArgs)
    error = new ConsoleError(formattedErrorMessage)
  } else {
    error = originError
  }

  storeHydrationErrorStateFromConsoleArgs(...consoleErrorArgs)
  attachHydrationErrorState(error)

  const componentStackTrace =
    (error as any)._componentStack || hydrationErrorState.componentStack
  const componentStackFrames =
    typeof componentStackTrace === 'string'
      ? parseComponentStack(componentStackTrace)
      : undefined

  // Skip ModuleBuildError and ModuleNotFoundError, as it will be sent through onBuildError callback.
  // This is to avoid same error as different type showing up on client to cause flashing.
  if (
    error.name !== 'ModuleBuildError' &&
    error.name !== 'ModuleNotFoundError'
  ) {
    Bus.emit({
      type: ACTION_UNHANDLED_ERROR,
      reason: error,
      frames: parseStack(error.stack),
      componentStackFrames,
    })
  }
}

function onUnhandledError(event: ErrorEvent) {
  const error = event?.error
  handleClientError(error, [])
}

function onUnhandledRejection(ev: PromiseRejectionEvent) {
  const reason = ev?.reason
  if (
    !reason ||
    !(reason instanceof Error) ||
    typeof reason.stack !== 'string'
  ) {
    // A non-error was thrown, we don't have anything to show. :-(
    return
  }

  const e = reason
  Bus.emit({
    type: ACTION_UNHANDLED_REJECTION,
    reason: reason,
    frames: parseStack(e.stack!),
  })
}

export function register() {
  if (isRegistered) {
    return
  }
  isRegistered = true

  try {
    const limit = Error.stackTraceLimit
    Error.stackTraceLimit = 50
    stackTraceLimit = limit
  } catch {}

  window.addEventListener('error', onUnhandledError)
  window.addEventListener('unhandledrejection', onUnhandledRejection)
  patchConsoleError(handleClientError)
}

export function unregister() {
  if (!isRegistered) {
    return
  }
  isRegistered = false

  if (stackTraceLimit !== undefined) {
    try {
      Error.stackTraceLimit = stackTraceLimit
    } catch {}
    stackTraceLimit = undefined
  }

  window.removeEventListener('error', onUnhandledError)
  window.removeEventListener('unhandledrejection', onUnhandledRejection)
  unpatchConsoleError()
}

export function onBuildOk() {
  Bus.emit({ type: ACTION_BUILD_OK })
}

export function onBuildError(message: string) {
  Bus.emit({ type: ACTION_BUILD_ERROR, message })
}

export function onRefresh() {
  Bus.emit({ type: ACTION_REFRESH })
}

export function onBeforeRefresh() {
  Bus.emit({ type: ACTION_BEFORE_REFRESH })
}

export function onVersionInfo(versionInfo: VersionInfo) {
  Bus.emit({ type: ACTION_VERSION_INFO, versionInfo })
}

export { getErrorByType } from '../internal/helpers/get-error-by-type'
export { getServerError } from '../internal/helpers/node-stack-frames'
export { default as ReactDevOverlay } from './ReactDevOverlay'
