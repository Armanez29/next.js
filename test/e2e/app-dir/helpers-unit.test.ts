import { createOptimisticTree } from 'next/client/components/helpers'
import { AppRouterState } from 'next/client/components/app-router.client'

describe('Routing helpers', () => {
  it('should create the correct optimistic tree for same-layout navigation', () => {
    const existingState: AppRouterState = {
      tree: [
        '',
        {
          children: [
            'dashboard',
            {
              children: ['page', {}],
            },
          ],
        },
        '/dashboard',
      ],
      cache: {
        subTreeData: null,
        childNodes: new Map([
          [
            'dashboard',
            {
              subTreeData: [
                {
                  type: 'h1',
                  key: null,
                  ref: null,
                  props: {
                    children: 'Dashboard',
                  },
                  _owner: null,
                  _store: {},
                },
                {
                  type: {
                    _payload: {
                      _status: 3,
                      _response: {
                        _bundlerConfig: null,
                        _chunks: {},
                        _partialRow: '',
                        _stringDecoder: {},
                      },
                    },
                  },
                  key: null,
                  ref: null,
                  props: {
                    layoutPath: '/dashboard',
                    childProp: {
                      current: null,
                      segment: 'page',
                    },
                  },
                  _owner: null,
                  _store: {},
                },
              ],
              childNodes: new Map([
                [
                  'page',
                  {
                    subTreeData: [
                      {
                        type: 'h1',
                        key: null,
                        ref: null,
                        props: {
                          children: 'Dashboard',
                        },
                        _owner: null,
                        _store: {},
                      },
                      {
                        type: {
                          _payload: {
                            _status: 3,
                            _response: {
                              _bundlerConfig: null,
                              _chunks: {},
                              _partialRow: '',
                              _stringDecoder: {},
                            },
                          },
                        },
                        key: null,
                        ref: null,
                        props: {
                          layoutPath: '/dashboard',
                          childProp: {
                            current: null,
                            segment: 'page',
                          },
                        },
                        _owner: null,
                        _store: {},
                      },
                    ],
                    childNodes: new Map(),
                  },
                ],
              ]),
            },
          ],
        ]),
      },
    }

    const fetch = global.fetch
    global.fetch = jest.fn(() => new Promise(() => {}))

    const href = '/dashboard/integrations'
    const url = new URL(href, 'http://localhost:3000')
    const segments = url.pathname.split('/')
    segments.push('page')
    const optimisticTree = createOptimisticTree(
      existingState,
      url,
      segments,
      existingState.tree,
      existingState.cache.childNodes,
      true,
      false
    )

    expect(existingState).toMatchSnapshot()

    expect(optimisticTree).toMatchSnapshot()

    global.fetch = fetch
  })

  it('should create the correct optimistic tree for different-layout navigation', () => {
    const existingState: AppRouterState = {
      tree: [
        '',
        {
          children: [
            'dashboard',
            {
              children: ['page', {}],
            },
          ],
        },
        '/dashboard',
      ],
      cache: {
        subTreeData: null,
        childNodes: new Map([
          [
            'dashboard',
            {
              subTreeData: [
                {
                  type: 'h1',
                  key: null,
                  ref: null,
                  props: {
                    children: 'Dashboard',
                  },
                  _owner: null,
                  _store: {},
                },
                {
                  type: {
                    _payload: {
                      _status: 3,
                      _response: {
                        _bundlerConfig: null,
                        _chunks: {},
                        _partialRow: '',
                        _stringDecoder: {},
                      },
                    },
                  },
                  key: null,
                  ref: null,
                  props: {
                    layoutPath: '/dashboard',
                    childProp: {
                      current: null,
                      segment: 'page',
                    },
                  },
                  _owner: null,
                  _store: {},
                },
              ],
              childNodes: new Map([
                [
                  'page',
                  {
                    subTreeData: [
                      {
                        type: 'h1',
                        key: null,
                        ref: null,
                        props: {
                          children: 'Dashboard',
                        },
                        _owner: null,
                        _store: {},
                      },
                      {
                        type: {
                          _payload: {
                            _status: 3,
                            _response: {
                              _bundlerConfig: null,
                              _chunks: {},
                              _partialRow: '',
                              _stringDecoder: {},
                            },
                          },
                        },
                        key: null,
                        ref: null,
                        props: {
                          layoutPath: '/dashboard',
                          childProp: {
                            current: null,
                            segment: 'page',
                          },
                        },
                        _owner: null,
                        _store: {},
                      },
                    ],
                    childNodes: new Map(),
                  },
                ],
              ]),
            },
          ],
        ]),
      },
    }

    const fetch = global.fetch
    global.fetch = jest.fn(() => new Promise(() => {}))

    const href = '/slow-page-with-loading'
    const url = new URL(href, 'http://localhost:3000')
    const segments = url.pathname.split('/')
    segments.push('page')
    const optimisticTree = createOptimisticTree(
      existingState,
      url,
      segments,
      existingState.tree,
      existingState.cache.childNodes,
      true,
      false
    )

    expect(existingState).toMatchSnapshot()

    expect(optimisticTree).toMatchSnapshot()

    global.fetch = fetch
  })
})
