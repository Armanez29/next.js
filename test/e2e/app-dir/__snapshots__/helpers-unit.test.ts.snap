// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routing helpers should create the correct optimistic tree for different-layout navigation 1`] = `
Object {
  "cache": Object {
    "childNodes": Map {
      "dashboard" => Object {
        "childNodes": Map {
          "page" => Object {
            "childNodes": Map {},
            "subTreeData": Array [
              Object {
                "_owner": null,
                "_store": Object {},
                "key": null,
                "props": Object {
                  "children": "Dashboard",
                },
                "ref": null,
                "type": "h1",
              },
              Object {
                "_owner": null,
                "_store": Object {},
                "key": null,
                "props": Object {
                  "childProp": Object {
                    "current": null,
                    "segment": "page",
                  },
                  "layoutPath": "/dashboard",
                },
                "ref": null,
                "type": Object {
                  "_payload": Object {
                    "_response": Object {
                      "_bundlerConfig": null,
                      "_chunks": Object {},
                      "_partialRow": "",
                      "_stringDecoder": Object {},
                    },
                    "_status": 3,
                  },
                },
              },
            ],
          },
        },
        "subTreeData": Array [
          Object {
            "_owner": null,
            "_store": Object {},
            "key": null,
            "props": Object {
              "children": "Dashboard",
            },
            "ref": null,
            "type": "h1",
          },
          Object {
            "_owner": null,
            "_store": Object {},
            "key": null,
            "props": Object {
              "childProp": Object {
                "current": null,
                "segment": "page",
              },
              "layoutPath": "/dashboard",
            },
            "ref": null,
            "type": Object {
              "_payload": Object {
                "_response": Object {
                  "_bundlerConfig": null,
                  "_chunks": Object {},
                  "_partialRow": "",
                  "_stringDecoder": Object {},
                },
                "_status": 3,
              },
            },
          },
        ],
      },
    },
    "subTreeData": null,
  },
  "tree": Array [
    "",
    Object {
      "children": Array [
        "dashboard",
        Object {
          "children": Array [
            "page",
            Object {},
          ],
        },
      ],
    },
    "/dashboard",
  ],
}
`;

exports[`Routing helpers should create the correct optimistic tree for different-layout navigation 2`] = `
Object {
  "childNodes": Map {
    "dashboard" => Object {
      "childNodes": Map {
        "page" => Object {
          "childNodes": Map {},
          "subTreeData": Array [
            Object {
              "_owner": null,
              "_store": Object {},
              "key": null,
              "props": Object {
                "children": "Dashboard",
              },
              "ref": null,
              "type": "h1",
            },
            Object {
              "_owner": null,
              "_store": Object {},
              "key": null,
              "props": Object {
                "childProp": Object {
                  "current": null,
                  "segment": "page",
                },
                "layoutPath": "/dashboard",
              },
              "ref": null,
              "type": Object {
                "_payload": Object {
                  "_response": Object {
                    "_bundlerConfig": null,
                    "_chunks": Object {},
                    "_partialRow": "",
                    "_stringDecoder": Object {},
                  },
                  "_status": 3,
                },
              },
            },
          ],
        },
      },
      "subTreeData": Array [
        Object {
          "_owner": null,
          "_store": Object {},
          "key": null,
          "props": Object {
            "children": "Dashboard",
          },
          "ref": null,
          "type": "h1",
        },
        Object {
          "_owner": null,
          "_store": Object {},
          "key": null,
          "props": Object {
            "childProp": Object {
              "current": null,
              "segment": "page",
            },
            "layoutPath": "/dashboard",
          },
          "ref": null,
          "type": Object {
            "_payload": Object {
              "_response": Object {
                "_bundlerConfig": null,
                "_chunks": Object {},
                "_partialRow": "",
                "_stringDecoder": Object {},
              },
              "_status": 3,
            },
          },
        },
      ],
    },
  },
  "tree": Array [
    "",
    Object {
      "children": Array [
        "slow-page-with-loading",
        Object {
          "children": Array [
            "page",
            Object {},
          ],
        },
      ],
    },
    "/slow-page-with-loading",
  ],
}
`;

exports[`Routing helpers should create the correct optimistic tree for same-layout navigation 1`] = `
Object {
  "cache": Object {
    "childNodes": Map {
      "dashboard" => Object {
        "childNodes": Map {
          "page" => Object {
            "childNodes": Map {},
            "subTreeData": Array [
              Object {
                "_owner": null,
                "_store": Object {},
                "key": null,
                "props": Object {
                  "children": "Dashboard",
                },
                "ref": null,
                "type": "h1",
              },
              Object {
                "_owner": null,
                "_store": Object {},
                "key": null,
                "props": Object {
                  "childProp": Object {
                    "current": null,
                    "segment": "page",
                  },
                  "layoutPath": "/dashboard",
                },
                "ref": null,
                "type": Object {
                  "_payload": Object {
                    "_response": Object {
                      "_bundlerConfig": null,
                      "_chunks": Object {},
                      "_partialRow": "",
                      "_stringDecoder": Object {},
                    },
                    "_status": 3,
                  },
                },
              },
            ],
          },
        },
        "subTreeData": Array [
          Object {
            "_owner": null,
            "_store": Object {},
            "key": null,
            "props": Object {
              "children": "Dashboard",
            },
            "ref": null,
            "type": "h1",
          },
          Object {
            "_owner": null,
            "_store": Object {},
            "key": null,
            "props": Object {
              "childProp": Object {
                "current": null,
                "segment": "page",
              },
              "layoutPath": "/dashboard",
            },
            "ref": null,
            "type": Object {
              "_payload": Object {
                "_response": Object {
                  "_bundlerConfig": null,
                  "_chunks": Object {},
                  "_partialRow": "",
                  "_stringDecoder": Object {},
                },
                "_status": 3,
              },
            },
          },
        ],
      },
    },
    "subTreeData": null,
  },
  "tree": Array [
    "",
    Object {
      "children": Array [
        "dashboard",
        Object {
          "children": Array [
            "page",
            Object {},
          ],
        },
      ],
    },
    "/dashboard",
  ],
}
`;

exports[`Routing helpers should create the correct optimistic tree for same-layout navigation 2`] = `
Object {
  "childNodes": Map {
    "dashboard" => Object {
      "childNodes": Map {
        "page" => Object {
          "childNodes": Map {},
          "subTreeData": Array [
            Object {
              "_owner": null,
              "_store": Object {},
              "key": null,
              "props": Object {
                "children": "Dashboard",
              },
              "ref": null,
              "type": "h1",
            },
            Object {
              "_owner": null,
              "_store": Object {},
              "key": null,
              "props": Object {
                "childProp": Object {
                  "current": null,
                  "segment": "page",
                },
                "layoutPath": "/dashboard",
              },
              "ref": null,
              "type": Object {
                "_payload": Object {
                  "_response": Object {
                    "_bundlerConfig": null,
                    "_chunks": Object {},
                    "_partialRow": "",
                    "_stringDecoder": Object {},
                  },
                  "_status": 3,
                },
              },
            },
          ],
        },
      },
      "subTreeData": Array [
        Object {
          "_owner": null,
          "_store": Object {},
          "key": null,
          "props": Object {
            "children": "Dashboard",
          },
          "ref": null,
          "type": "h1",
        },
        Object {
          "_owner": null,
          "_store": Object {},
          "key": null,
          "props": Object {
            "childProp": Object {
              "current": null,
              "segment": "page",
            },
            "layoutPath": "/dashboard",
          },
          "ref": null,
          "type": Object {
            "_payload": Object {
              "_response": Object {
                "_bundlerConfig": null,
                "_chunks": Object {},
                "_partialRow": "",
                "_stringDecoder": Object {},
              },
              "_status": 3,
            },
          },
        },
      ],
    },
  },
  "tree": Array [
    "",
    Object {
      "children": Array [
        "dashboard",
        Object {
          "children": Array [
            "integrations",
            Object {
              "children": Array [
                "page",
                Object {},
              ],
            },
          ],
        },
      ],
    },
    "/dashboard/integrations",
  ],
}
`;
